x/retry/config.go

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package retry
+
+import (
+	"fmt"
+	"time"
+
+	"go.uber.org/multierr"
+	iconfig "go.uber.org/yarpc/internal/config"
+	"go.uber.org/yarpc/x/config"
+)
+
+// PolicySpec defines how to construct a retry Policy.
+type PolicySpec struct {
+	// Retries indicates the number of retries that will be attempted on failed
+	// requests.
+	Retries uint `config:"retries"`
+
+	// MaxTimeout indicates the max timeout that every request through the retry
+	// middleware will have.  If the timeout is greater than the request timeout
+	// we'll use the request timeout instead.
+	MaxTimeout time.Duration `config:"maxTimeout"`
+
+	// BackoffStrategy defines a backoff strategy in place by embedding a
+	// backoff config.
+	BackoffStrategy config.Backoff `config:"backoff"`
+}
+
+func (p PolicySpec) policy() (*Policy, error) {
+	strategy, err := p.BackoffStrategy.Strategy()
+	if err != nil {
+		return nil, err
+	}
+	return NewPolicy(
+		Retries(p.Retries),
+		MaxRequestTimeout(p.MaxTimeout),
+		BackoffStrategy(strategy),
+	), nil
+}
+
+// PolicyOverrideSpec defines per service or per service+procedure Policies that
+// will be applied in the PolicyProvider.
+type PolicyOverrideSpec struct {
```

Comment:
Same comment as above. Also, for structs representing configuration, you should have examples in the docs.